### PR TITLE
Gitlab: filter untouched pkgs from pipeline

### PR DESCRIPTION
**Pull request by scottwittenburg**
We've previously generated CI pipelines for PRs, and they rebuild any packages that don't have a binary in an existing build cache.  The assumption we were making was that ALL prior merged builds would be in cache, but due to the way we do security in the pipeline, they aren't. `develop` pipelines can take a while to catch up with the latest PRs, and while it does that, there may be a bunch of redundant builds on PRs that duplicate things being rebuilt on `develop`.  Until we can do better caching of PR builds, we'll have this problem.

We can do better in PRs, though, by *only* rebuilding things in the CI environment that are actually touched by the PR.  This change computes exactly what packages are changed by a PR branch and *only* includes those packages' dependents and dependencies in the generated pipeline.  Other as-yet unbuilt packages are pruned from CI for the PR.

For `develop` pipelines, we still want to build everything to ensure that the stack works, and to ensure that `develop` catches up with PRs. This is especially true since we do not do rebuilds for *every* commit on `develop` -- just the most recent one after each `develop` pipeline finishes.  Since we skip around, we may end up missing builds unless we ensure that we rebuild everything.

We differentiate between `develop` and PR pipelines in `.gitlab-ci.yml` by setting `SPACK_PRUNE_UNTOUCHED` for PRs. `develop` will still have the old behavior.

- [x] Add `SPACK_PRUNE_UNTOUCHED` variable to `spack ci`
- [x] Refactor `spack pkg` command by moving historical package checking logic to `spack.repo`
- [x] Implement pruning logic in `spack ci` to remove untouched packages
- [x] add tests

### CODE REVIEW

**Comment by scottwittenburg** [comment]
I think I need to change the fetch depth in gitlab to at least two before this has a chance there.

File: lib/spack/spack/ci.py
Comment (trws) [inline lib/spack/spack/ci.py:524]: Perhaps `git merge-base HEAD develop`? Better would be to have that be default and allow passing it in, then use the PR target branch in CI, that'll always get the whole thing.

Comment (tgamblin) [reply lib/spack/spack/ci.py:524]: possibly better b/c you can still do a shallow clone: use the GitHub API.  See https://github.com/spack/spackbot/blob/main/spackbot/handlers/labels.py#L137-L175

See docs here: https://developer.github.com/v3/pulls/#list-pull-requests-files

Comment (scottwittenburg) [reply lib/spack/spack/ci.py:524]: Can we get another identity to use for these api requests?  I ask because GitHub is already rate-limiting the one we have for some part of most hours.  @zackgalbreath 

Otherwise, yeah, we can do a full clone and get the merge-base between the parents of the merge commit we're testing.  Another suggestion (thanks to @opadron) was we just set the fetch depth to some "reasonable" value, and if we can't find the merge-base for PRs, print a message suggesting rebase.

I'm doing some experiments with all the branches I cloned from gitlab yesterday to see what "reasonable" could look like for that.

Comment (tgamblin) [reply lib/spack/spack/ci.py:524]: Actions have their own GITHUB_TOKEN that you can use for API calls. It won’t have permission to write to the repo or to add users to teams like spackbot needs but it can query things like files modified in a PR.

Comment (tgamblin) [reply lib/spack/spack/ci.py:524]: I guess since this is running in GitLab not GitHub I’m not sure you actually get that token though… hm.

Comment (tgamblin) [reply lib/spack/spack/ci.py:524]: Here we go — you can do it with the GitLab API: https://docs.gitlab.com/ee/api/repositories.html#compare-branches-tags-or-commits

there should be a token for GitLab in the CI env that could be used to talk back to gitlab.spack.io

Comment (scottwittenburg) [reply lib/spack/spack/ci.py:524]: For PR testing where we test the merge commit from github, my opinion is we want the values in this PR now (compare `HEAD^` and `HEAD`).  I wrote a little python script to check all the branches I pulled yesterday, and compare the changed files using the current approach against the changed files as returned by the github api.  Of the 498 branches my script checked, there were 14 where the file list didn't match and I looked at those PRs on github and found that they all fell into one of the following categories:

- waiting for style check to pass
- conflicts and cannot be merged
- draft PR
- branch pushed again between when I cloned the repo and ran the script

The first three conditions are noticed by the sync script and it avoids pushing in those cases.

This makes sense to me since I'd expect the merge commit made by github and returned to us in the gh api request for a PR to be consistent with the gh api request for the changed files of that same PR.

Comment (scottwittenburg) [reply lib/spack/spack/ci.py:524]: But I agree with @trws that it would be better to have a reasonable default and allow passing it in.

**Comment by scottwittenburg** [comment]
I force pushed without changes to see if it unsticks all the checks.  I'd like to see the gitlab pipeline build nothing, then I'll re-write the history to have a few commits with packages changes at the beginning and make sure those things (along with the things in their cones of destruction) get scheduled.

**Comment by scottwittenburg** [comment]
Pipeline scheduled no jobs when no pkgs were touched: [108659](https://gitlab.spack.io/spack/spack/-/pipelines/108659)

File: lib/spack/spack/cmd/pkg.py
Comment (alalazo) [inline lib/spack/spack/cmd/pkg.py]: Can we move this function somewhere in "core" Spack, i.e. not as part of a `spack.cmd.`? In that way `spack.ci` doesn't need to depend on a `spack.cmd` package.

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py]: We can avoid having Spack "core" API depend on `spack.cmd` if we move the function we need from `pkg_cmd` to another module.

**Review by alalazo — COMMENTED** [review]
A minor comment on code organization.

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py]: Would this work too:
```suggestion
        env_matches = [s for s in all_concrete_specs if s.name == pkg]
```
?

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py]: I think we can do:
```suggestion
                affected_specs.update(match.traverse(direction='children', root=False))
```
and similarly below.

File: lib/spack/spack/cmd/pkg.py
Comment (scottwittenburg) [reply lib/spack/spack/cmd/pkg.py]: Sure, any suggestions?  Otherwise, I'll look around and come up with a location.

File: lib/spack/spack/ci.py
Comment (scottwittenburg) [reply lib/spack/spack/ci.py]: I assume that would work.  Is the idea that it should be faster to do that way?

File: lib/spack/spack/ci.py
Comment (scottwittenburg) [reply lib/spack/spack/ci.py]: Sure, will do.  Is it about performance or code style (or number of lines)?

File: lib/spack/spack/ci.py
Comment (trws) [reply lib/spack/spack/ci.py:524]: If we do it that way, different CI providers can provide it however makes sense.  In an action use the GH api, in GL use the GL api or something else.  Another option for dealing with the slow checkouts, that I was looking at for the other tests because many (**MANY**) unit tests fail without a significant chunk of a clone for some reason, is to switch to using `--shallow-since=year/mm/dd` with the day before the PR began.  It's guaranteed to have everything the PR touches, uses only the git API, and cuts down the spack repo for a PR like this one to only ~10mb.

```console
❯ g clone --shallow-since=2022/03/22 https://github.com/spack/spack spack-shallow
Cloning into 'spack-shallow'...
remote: Enumerating objects: 16406, done.
remote: Counting objects: 100% (16406/16406), done.
remote: Compressing objects: 100% (9358/9358), done.
remote: Total 16406 (delta 1293), reused 9904 (delta 959), pack-reused 0
Receiving objects: 100% (16406/16406), 9.79 MiB | 13.84 MiB/s, done.
Resolving deltas: 100% (1293/1293), done.
```

The downside is that it may have to be cloned from upstream, then fetch in the changes from a fork, the forks don't always allow that kind of shallow query.

File: lib/spack/spack/ci.py
Comment (alalazo) [reply lib/spack/spack/ci.py]: Yes, that will be faster (it's a string comparison) and avoid Spec construction - not sure if it will be relevant overall but since we can avoid useless computation and still have clear code...

Comment (scottwittenburg) [reply lib/spack/spack/ci.py]: Oh yeah, I missed we could skip spec construction too.  Good catch, thanks :+1:

File: lib/spack/spack/ci.py
Comment (alalazo) [reply lib/spack/spack/ci.py]: Mainly personal preference on code style, not sure about performance but it won't be worse I guess. It's the "set" version of using:
```python
list.extend(sequence)
```
instead of:
```python
for item in sequence:
   list.append(item)
```

File: lib/spack/spack/cmd/pkg.py
Comment (scottwittenburg) [reply lib/spack/spack/cmd/pkg.py]: What about `lib/spack/spack/package.py`?  I would probably need to move some other methods in there as well, including `get_git`, and I'm not sure how good a fit that is.  There's also `lib/spack/spack/pkgkit.py` which currently seems to just collect stuff from other modules, but could maybe include a few module level methods.

Comment (scottwittenburg) [reply lib/spack/spack/cmd/pkg.py]: Oh, looking more, maybe not `pkgkit.py`, since it's goal is to make things available to package files.

Comment (trws) [reply lib/spack/spack/cmd/pkg.py]: Definitely not pkgkit, package could make sense, or perhaps repo since it's an "all packages" action?

**Comment by scottwittenburg** [comment]
Still looking for a good home for those `cmd/pkg.py` methods, but wanted to push again to see if it unsticks all the checks, so we can see touched things and their DAGs getting scheduled.

File: lib/spack/spack/cmd/pkg.py
Comment (alalazo) [reply lib/spack/spack/cmd/pkg.py]: I think we can put it under `spack.ci`? After all it tells, given 2 revisions, which packages changed in between. Where would that be useful if not in CI?

Comment (trws) [reply lib/spack/spack/cmd/pkg.py]: In unit-test, style (which actually has another version of the same functionality already IIRC), probably other places too.

Comment (scottwittenburg) [reply lib/spack/spack/cmd/pkg.py]: Additionally, `cmd/pkg.py` already depends on `spack.repo`, so I kind of like it for that reason.

Comment (alalazo) [reply lib/spack/spack/cmd/pkg.py]: Just wanted to note that unit-test and style are always run... in ci :laughing: Besides jokes the main point would be to avoid commands depending on "core" APIs and vice-versa so any core module mentioned here is fine with me for the time being.

**Comment by scottwittenburg** [comment]
It's hard to be 100% sure, but it looks like what got scheduled in this [pipeline](https://gitlab.spack.io/spack/spack/-/pipelines/108745) is just the intersection of:

1. what was touched (`py-beniget`) and its dual cones of destruction, and
2. what was erroneously scheduled (thanks to out of date binary index) under the e4s stack

So to me, it looks like this is working as expected.

File: lib/spack/spack/cmd/pkg.py
Comment (scottwittenburg) [reply lib/spack/spack/cmd/pkg.py]: Ok, I had to move all the common methods, so they're living in `lib/spack/spack/repo.py` now.

**Comment by scottwittenburg** [comment]
I want to leave the `py-beniget` change in for one more round of ci to make sure the refactor didn't break anything (I already tested locally, but still).  If gitlab schedules the expected jobs, I'll remove that WIP commit, and then this could be good to go.

Associated pipeline: [108855](https://gitlab.spack.io/spack/spack/-/pipelines/108855)

**Comment by alalazo** [comment]
Are tests stateful in `ci.py`? It seems we added unit tests and an unrelated test in the same file is failing consistently.

**Comment by scottwittenburg** [comment]
> Are tests stateful in `ci.py`? It seems we added unit tests and an unrelated test in the same file is failing consistently.

Yes, that was due to setting an environment variable and forgetting to unset it after it was needed, so the next tests had it turned on unexpectedly.

File: lib/spack/spack/test/cmd/ci.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/ci.py:1499]: Got it, thanks @scottwittenburg For things like this you can use [monkeypatch.setenv](https://docs.pytest.org/en/6.2.x/_modules/_pytest/monkeypatch.html#MonkeyPatch.setenv) which restore the state automatically.

Comment (scottwittenburg) [reply lib/spack/spack/test/cmd/ci.py:1499]: Looking around I see there's also a `working_env` fixture defined [here](https://github.com/spack/spack/blob/36fcf461d934eacf9a33b59a9097ef8be9cf6d69/lib/spack/spack/test/conftest.py#L359), which seems to be used in a lot of places.

I created a draft PR (#29807) to make use of that fixture in the ci tests.  I'll rebase it on `develop` after this one is merged if that's ok with you.  I feel it would be good to get this in asap, and remove a lot of the unnecessary workload from gitlab.

Comment (alalazo) [reply lib/spack/spack/test/cmd/ci.py:1499]: Yeah I agree with that.

Comment (scottwittenburg) [reply lib/spack/spack/test/cmd/ci.py:1499]: Just needs an approving review.

File: lib/spack/spack/ci.py
Comment (trws) [reply lib/spack/spack/ci.py:524]: This is really my last concern, there are a good number of longer commit chains in PRs, is the plan to leave this TODO in place?

**Review by tgamblin — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

File: lib/spack/spack/ci.py
Comment (scottwittenburg) [reply lib/spack/spack/ci.py:524]: If your concern is about whether this is going to work for gitlab pipelines run on spack PRs, I took some steps (described [here](https://github.com/spack/spack/pull/29697#discussion_r834912192), [here](https://github.com/spack/spack/pull/29697#issuecomment-1078604570), and [here](https://github.com/spack/spack/pull/29697#issuecomment-1079112231)) to make sure this approach is going to be sufficient for our intended usage.

You mentioned the ability to do things differently if we're in a github action, but we're relatively far from being able to generate a pipeline that runs in github actions (or any other CI provider), so I'm hesitant to hold this up over that.  

I could do something super-quick like allow you to override those values with a couple of environment variables.  Would that be useful for something you have mind (and get you on board for merging this :laughing:)?  I'd be happy to do that, and I'd be really interested what it is you have in mind.

Also, I got the feeling @tgamblin wanted to reduce the number of jobs we're generating the past couple weeks, and that time was of the essence (I guess so, since he just merged it).

Comment (trws) [reply lib/spack/spack/ci.py:524]: Nah, if it's getting the difference between the PR and develop (however it's getting it) I'm good with that.  The only thing I think would really be practical if we need the state back to merge base and still want shallow clones is to use the time-based shallow clone option.  If we don't need that to get the list of everything changed between the PR and develop then I wouldn't worry about it at all at this point.

Comment (tgamblin) [reply lib/spack/spack/ci.py:524]: My understanding is we do not need it because we clone with depth 2 from the merge commit that's built for each PR.  On PR branches, that is all changes on the PR.  If we were to use this on `develop` or on some branch, yeah it'd just look at the difference with the last commit, but IMO that's the only thing that could make sense with pruning -- look at first parent.

But we wouldn't use pruning on `develop` or some branch -- we'll be using the old method there, where we just rebuild what isn't in cache.  Since the goal on a branch like `develop` is really to populate the cache, I don't see that as a problem. It's on the PRs where we get redundant rebuilds and want to reduce the builds.

Comment (trws) [reply lib/spack/spack/ci.py:524]: Yup, in case my wording was unclear I completely agree with that.  As long as
the PR builds are getting all the changes in the PR I don't see a problem.
Still might be nice to be able to specify it for testing outside automated CI,
but that's by no means pressing.

**Pull request merged**
